include/Mesh.h

```diff
@@ -301,14 +382,104 @@ template<typename real_t> class Mesh
     {
         // Sweep through boundary and set ids.
         size_t NElements = get_number_elements();
-	boundary.resize(NElements*nloc);
+        boundary.resize(NElements*nloc);
+        max_bdry_tag = 0;
         for(int i=0; i<NElements; i++) {
             for(int j=0; j<nloc; j++) {
-                boundary[i*nloc+j] = _boundary[i*nloc+j];
+                int b = _boundary[i*nloc+j];
+                boundary[i*nloc+j] = b;
+                if (b > max_bdry_tag)
+                    max_bdry_tag = b;
+            }
+        }
+    }
+
+    void set_regions(const int *_regions)
+    {
+        // Sweep through elements and set ids.
+        size_t NElements = get_number_elements();
+        regions.resize(NElements);
+        if (!_regions) {
+            std::fill(regions.begin(), regions.end(), 0);
+        }
+        else {
+            memcpy(&regions[0], _regions, NElements*sizeof(int));
+        }
+    }
+
+    ///  add internal boundary tags if facet is between 2+ regions
+    ///   unless they were already provided ? TODO check consistency
```

Comment:
yes please check !

Reply:
It is consistent if we suppose we don't preserve internal boundaries and they cannot be prescribed.